### PR TITLE
Feature branch

### PR DESCRIPTION
**Pull request by amita-roy**
## Private Events

> We have created 3 models, User, Event and EventAttendance
> User is associated with event as a creator
> User is associated with the event as an attendee and attended_event
> We can see all events on root url but only a logged in user can create a new event
> We have 3 tabs on navbar, All events, Past and Upcoming
> On user show page we can see Events that user has created and the events that user is going to attend (Upcoming) or attended(Past).

> On Event show page, we can see event details along with the attendees list

### CODE REVIEW

File: app/models/event.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class Event < ApplicationRecord
+  belongs_to :creator, class_name: :User, foreign_key: :creator_id
```
Comment (mcrd25) [inline app/models/event.rb]: `foreign_key: :creator_id` not necessary here

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      redirect_to @user, notice: 'Thanks for signing up!'
+    else
+      render :new
+    end
+  end
+
+  def show
+    @user = User.find(params[:id])
+    @past_events = @user.attended_events.where('event_date < ?', Time.now).order('event_date DESC')
+    @upcoming_events = @user.attended_events.where('event_date > ?', Time.now).order('event_date DESC')
```
Comment (mcrd25) [inline app/controllers/users_controller.rb:23]: User scopes created in your event model here.

File: app/models/event.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+  has_many :event_attendances, foreign_key: :attended_event_id
+  has_many :attendees, through: :event_attendances, source: :attendee
+  has_one_attached :main_image
+
+  scope :past, -> { where('event_date < ?', Time.now).order('event_date DESC') }
+  scope :upcoming, -> { where('event_date > ?', Time.now).order('event_date DESC') }
```
Comment (mcrd25) [inline app/models/event.rb:11]: Use these scopes in your `@past_events` and `@upcoming_events` in your Event Controller's show action.

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Changes Required :mag: 
Good job so far! :slightly_smiling_face: 
Your codce seems ok, with the exception of a bug that causes the following error:
Screenshot from 2020-10-22 11-15-04
The reviewer would better know if there is anything else left to change once this bug is fixed.
Please also address Review comments before submitting for rereview.

### Required Changes 
- Sign up page bug
- Review comments not marked as [OPTIONAL]
- You are missing your unit and feature tests using RSpec as well! See these screenshots from pathwright. [Unit Tests Requirement](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) and [Integration tests requirement](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png)


Otherwise, good job!


***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

*Happy coding 😄,*
**[--Maya Douglas](mayacdouglas.com)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

File: app/models/event.rb (hunk 1 above)
Comment (amita-roy) [reply app/models/event.rb]: removed the foreign_key: :creator_id  as suggested

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (amita-roy) [reply app/controllers/users_controller.rb:23]: There are the scopes for user's attended_events (Past, Upcoming) which you can see on User show page. 


Screenshot 2020-10-23 at 8 52 20 PM

File: app/models/event.rb (hunk 3 above)
Comment (amita-roy) [reply app/models/event.rb:11]: These scopes are created for Past events and Upcoming event for Event Model. Which we can navigate through navbar tabs.

Screenshot 2020-10-22 at 6 44 09 PM

Screenshot 2020-10-22 at 6 44 01 PM

**Comment by amita-roy** [comment]
> Sign up page bug has been corrected

> Review comments has been corrected and responded

> Unit test and Integration test has been added

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi **TEAM**, great job on this project :+1: Nice job fixing all the comments made by the previous TSE :slightly_smiling_face: I particularly liked how you did the filter method with the event scopes :fireworks: awesome. However, we need to implement one feature which is currently missing. Kindly check the comments below:
- There is currently no button to activate the attend event. The user should be able to choose to attend the event or not 
- Also let us add this command `rails active_storage:install` to the README file. I needed to run that

Let us kindly implement this feature
Once we are done, we can submit another code review :smile: 
_Happy coding_ :woman_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Comment by amita-roy** [comment]
As per the last review

> We have added a button on event detail page to attend or un-attend the event.
> Only logged in user can attend the event
> User can only attend the events which are current dated or future dated
> If an event is past dated there will be no option to attend.

> This command 'rails active_storage:install' has been added to the README file as per the review.

**Review by rOluochKe — APPROVED** [review]
Hello Team
### You got it right, well done :clap: The project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one. 
- Happy coding!

**Pull request merged**